k8sutils/pkg/instrumentation_instance/status.go

```
@@ -63,43 +72,70 @@ func InstrumentationInstanceName(ownerName string, pid int) string {
 func UpdateInstrumentationInstanceStatus(ctx context.Context, owner client.Object, containerName string, kubeClient client.Client,
 	instrumentedAppName string, pid int, scheme *runtime.Scheme, options ...InstrumentationInstanceOption) error {
 	instrumentationInstanceName := InstrumentationInstanceName(owner.GetName(), pid)
-	updatedInstance := &odigosv1.InstrumentationInstance{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: "odigos.io/v1alpha1",
-			Kind:       "InstrumentationInstance",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      instrumentationInstanceName,
-			Namespace: owner.GetNamespace(),
-			Labels: map[string]string{
-				consts.InstrumentedAppNameLabel: instrumentedAppName,
-			},
-		},
-		Spec: odigosv1.InstrumentationInstanceSpec{
-			ContainerName: containerName,
-		},
-	}
+	instance := odigosv1.InstrumentationInstance{}
 
-	err := controllerutil.SetControllerReference(owner, updatedInstance, scheme)
+	err := kubeClient.Get(ctx,
+		client.ObjectKey{Namespace: owner.GetNamespace(), Name: instrumentationInstanceName},
+		&instance,
+	)
 	if err != nil {
-		return err
-	}
-
-	if err = kubeClient.Create(ctx, updatedInstance); err != nil {
-		if !apierrors.IsAlreadyExists(err) {
+		if !apierrors.IsNotFound(err) {
 			return err
 		}
-		err := kubeClient.Get(ctx, client.ObjectKeyFromObject(updatedInstance), updatedInstance)
+
+		// check if we can create a new instance
+		if instrumentationInstancesCountReachedLimit(ctx, owner, kubeClient) {
+			return fmt.Errorf("instrumentation instances count per pod is over the limit of %d", maxInstrumentationInstancesPerPod)
```

I guess it's not really an error, but normal operation.
Will it show up in the logs as error with stack trace?
Perhaps we can just log as info here and return nil?

---

I prefer to return an error and let the caller decide what to do with it.
I can change that to an exported type `ErrInstanceReachedLimit` - wdyt?